### DISCUSSION TITLE
Why don't generated reports take uid from allure-results?

### DISCUSSION THREAD

**Discussion by noomorph** [allure-framework/allure2#2512] [Ideas & Feedback]
I am troubled by this inconsistency — I have no deterministic way to generate a link to a specific test case result in the HTML report, even if I know what the "id" is in the container or result JSON.

Is there a good reason?

![image](https://github.com/allure-framework/allure2/assets/1962469/854f270d-8f81-4b36-bdfc-f0f35e2bab16)

**Comment by noomorph** [comment]
In other words, why do you use `dest.setUid(uidGenerator.get());` here:

https://github.com/allure-framework/allure2/blob/main/allure-generator/src/main/java/io/qameta/allure/allure2/Allure2Plugin.java#L184

instead of taking `result.getUuid()`?

**Reply by baev** [reply]
1. The `uuid` field in the test result file is not required. 
2. The value isn't guaranteed to be unique or UUIDv4. It can be any string.
3. Using UUIDs in the report navigation increases the links' length. 

The current implementation was introduced to address the URL length problem. It uses random strings with just enough entropy to avoid collisions but much shorter. 

There are a few issues:

1. The test results `uids` are not stable. Having the report generated multiple times from the same result will produce different links for the same test results.
2. There is no way to predict a link for a particular test result during a test run via the Runtime API.
3. It is impossible to get a link for test results by 3rd party software by parsing results files.

We'll make the links stable in Allure 3. There is a chance it will also be backported to Allure 2 (depending on how much time we'll have). The rest of the issues are much lower priority.

**Reply by noomorph** [reply]
So, if I need it sooner, I may try luck with a fork, right?

Indeed, I need to report deep links in advance before the report is even generated. 🤔 
In some cases, post-processing `allure-report` could do the job as well, though.

Thanks for the explanation! 👍

**Reply by baev** [reply]
It would be nice if you bring it to light and share some insights on the problem you're trying to solve. 

Initially, the idea was to use custom plugins and generate the needed output with all the links from the report (or publish report links via API directly, similar to the Jira/Xray plugin we provide). So, instead of post-processing the existing `allure-report,` the custom plugin may provide you with a separate JSON with all the data needed.

**Reply by noomorph** [reply]
Okay, so to put it short, I wanted to have a guarantee that if I add a test like this:

```java
@AllureId('UNIQ-1`)
someTestCase() {
 // or whatever: allure.id('UNIQ-1') inside test body
}
```

that I will be able to find it on `https://allure-framework.github.io/allure-demo/5/#testresult/UNIQ-1`.

In my case, I'd like to embed iframes with a specific test result on a statically generated website. It is uncommon, but I just asked if it was possible. If I know the id in advance, I don't even need to regenerate the website — I can just regenerate allure-report if needed.

**Reply by baev** [reply]
The easiest way, in my opinion, would be to create a custom redirect layout in the report that would redirect `#allureId/UNIQ-1` to `#testresult/$uid`. To active that, all you need is to create a plugin that will create `data/redirects.json`, populate it with data needed (a simple map from-to will do), then create a layout that will access `data/redirects.json` and do the redirect if value found. 

By the way, using `@AllureId` for something else but Allure TestOps isn't recommended. Many features (including IDE integrations) may behave weirdly. Instead, you can create a custom label instead.

**Reply by noomorph** [reply]
Thank you, Dmitry. Looks like I have a clear picture of the status. Thanks for your time and effort to explain.

P.S. I hope you will find some time in the near future to review the Owner PR. My IntelliJ trial is expiring on 25th April 😅
